nat-lab/tests/test_vpn.py

```diff
@@ -253,3 +257,147 @@ async def test_vpn_reconnect(
+        ):
+            await client.connect_to_vpn(
+                wg_server["ipv4"], wg_server["port"], wg_server["public_key"]
+            )
+
+            async with Ping(connection, serv_ip).run() as ping:
+                await testing.wait_long(ping.wait_for_next_ping())
+
+        await connect(
+            config.WG_SERVER,
+        )
+
+        output_notifier = OutputNotifier()
+
+        async def on_stdout(stdout: str) -> None:
+            for line in stdout.splitlines():
+                output_notifier.handle_output(line)
+
+        async def conntrack_on_stdout(stdout: str) -> None:
+            for line in stdout.splitlines():
+                if f"dst={serv_ip}" in line:
+                    output_notifier.handle_output(line)
+
+        sender_start_event = asyncio.Event()
+        close_wait_event = asyncio.Event()
+
+        output_notifier.notify_output(
+            "80 port [tcp/*] succeeded!",
+            sender_start_event,
+        )
+        output_notifier.notify_output("CLOSE", close_wait_event)
+
+        async with connection.create_process(
+            [
+                "conntrack",
+                "--family",
+                "ipv6" if setup_params.ip_stack == IPStack.IPv6 else "ipv4",
+                "-E",
+            ]
+        ).run(stdout_callback=conntrack_on_stdout) as conntrack_proc:
+            await testing.wait_normal(conntrack_proc.wait_stdin_ready())
+
+            ip_proto = (
+                IPProto.IPv6 if setup_params.ip_stack == IPStack.IPv6 else IPProto.IPv4
+            )
+            alpha_ip = testing.unpack_optional(alpha.get_ip_address(ip_proto))
+
+            await exit_stack.enter_async_context(
+                connection.create_process(
+                    [
+                        "nc",
+                        "-nv",
+                        "-6" if setup_params.ip_stack == IPStack.IPv6 else "-4",
+                        "-s",
+                        alpha_ip,
+                        serv_ip,
+                        str(80),
+                    ]
+                ).run(stdout_callback=on_stdout, stderr_callback=on_stdout)
+            )
+
+            # Second client, this time sending some data to check proper TCP sequence number generation
+            proc = await exit_stack.enter_async_context(
+                connection.create_process(
+                    [
+                        "nc",
+                        "-nv",
+                        "-6" if setup_params.ip_stack == IPStack.IPv6 else "-4",
+                        "-s",
+                        alpha_ip,
+                        serv_ip,
+                        str(80),
+                    ]
+                ).run(stdout_callback=on_stdout, stderr_callback=on_stdout)
+            )
+
+            await proc.wait_stdin_ready()
+            await asyncio.sleep(2.0)
+            # Without this sleep nc get EOF on stdin for some reason
+            await proc.write_stdin("GET")
+
```

Comment:
non-functional: It is not necessary to push data through the TCP connection(as opposed to UDP) as it already did the 3-way handshake before it get's to the STDIN part at which point we already have all the building blocks for RST packet.

The existing code checks for more common case which is great but I felt the comment might be a bit misleading

Reply:
Adding the comment on the first client as well would be an improvement. Though all of this is minor you can feel free to ignore it